zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/AuthorizationService.java

```diff
@@ -0,0 +1,249 @@
+            isMember(entities, notebook.getNote(noteId).getOwners()) ||
+            isAdmin(entities);
+  }
+
+  public boolean isReader(String noteId, Set<String> entities) {
+    return isMember(entities, notebook.getNote(noteId).getReaders()) ||
+            isMember(entities, notebook.getNote(noteId).getOwners()) ||
+            isMember(entities, notebook.getNote(noteId).getWriters()) ||
+            isMember(entities, notebook.getNote(noteId).getRunners()) ||
+            isAdmin(entities);
+  }
+
+  public boolean isRunner(String noteId, Set<String> entities) {
+    return isMember(entities, notebook.getNote(noteId).getRunners()) ||
+            isMember(entities, notebook.getNote(noteId).getWriters()) ||
+            isMember(entities, notebook.getNote(noteId).getOwners()) ||
+            isAdmin(entities);
+  }
+
+  private boolean isAdmin(Set<String> entities) {
+    String adminRole = conf.getString(ZeppelinConfiguration.ConfVars.ZEPPELIN_OWNER_ROLE);
+    if (StringUtils.isBlank(adminRole)) {
+      return false;
+    }
+    return entities.contains(adminRole);
+  }
+
+  // return true if b is empty or if (a intersection b) is non-empty
+  private boolean isMember(Set<String> a, Set<String> b) {
+    Set<String> intersection = new HashSet<>(b);
+    intersection.retainAll(a);
+    return (b.isEmpty() || (intersection.size() > 0));
+  }
+
+  public boolean isOwner(Set<String> userAndRoles, String noteId) {
+    if (conf.isAnonymousAllowed()) {
+      LOGGER.debug("Zeppelin runs in anonymous mode, everybody is owner");
+      return true;
+    }
+    if (userAndRoles == null) {
+      return false;
+    }
+    return isOwner(noteId, userAndRoles);
+  }
+
+  //TODO(zjffdu) merge this hasWritePermission with isWriter ?
+  public boolean hasWritePermission(Set<String> userAndRoles, String noteId) {
+    if (conf.isAnonymousAllowed()) {
+      LOGGER.debug("Zeppelin runs in anonymous mode, everybody is writer");
+      return true;
+    }
+    if (userAndRoles == null) {
+      return false;
+    }
+    return isWriter(noteId, userAndRoles);
+  }
+
+  public boolean hasReadPermission(Set<String> userAndRoles, String noteId) {
+    if (conf.isAnonymousAllowed()) {
+      LOGGER.debug("Zeppelin runs in anonymous mode, everybody is reader");
+      return true;
+    }
+    if (userAndRoles == null) {
+      return false;
+    }
+    return isReader(noteId, userAndRoles);
+  }
+
+  public boolean hasRunPermission(Set<String> userAndRoles, String noteId) {
+    if (conf.isAnonymousAllowed()) {
+      LOGGER.debug("Zeppelin runs in anonymous mode, everybody is reader");
+      return true;
+    }
+    if (userAndRoles == null) {
+      return false;
+    }
+    return isReader(noteId, userAndRoles);
+  }
+
+  public boolean isPublic() {
+    return conf.isNotebookPublic();
+  }
+
+  public void setRoles(String user, Set<String> roles) {
+    if (StringUtils.isBlank(user)) {
+      LOGGER.warn("Setting roles for empty user");
```

Comment:
shouldn't this fail?

Reply:
This is from the old implementation: `NotebookAuthorization`, I guess it is to handle some corner case when user is empty.